# Unidad 2 Tarea 1 
### Fecha de entrega miércoles 14 de febrero del 2024. <a id="inicio"></a>
### Problema del PageRank 
Angel Mariel Osalde Salazar

# __Introducción__
<p style="text-align: justify;">
En estos ejercicios, exploraremos los conceptos fundamentales de los valores y vectores propios, los cuales desempeñan un papel crucial en el análisis de transformaciones lineales. Los valores propios son escalares que caracterizan cómo una transformación lineal cambia la dirección de ciertos vectores en el espacio, mientras que los vectores propios son aquellos asociados a estos valores y que conservan su dirección bajo la transformación.
</p>
<p style="text-align: justify;">
 Además, abordaremos el Modelo de Leslie, una herramienta matricial comúnmente utilizada en ecología para estudiar el crecimiento y la dinámica de las poblaciones en relación con sus diferentes grupos de edad. En este modelo, cada elemento de la matriz representa la tasa de supervivencia o reproducción de una clase de edad específica. A través de su análisis, esperamos obtener una comprensión más profunda de cómo este modelo influye en la evolución de una población particular.
</p>



# __Índice__

<ol>
    <li><a href="#int">Introducción</a></li>
    <li><a href="#Indice">Índice</a></li>
    <li><a href="#voyva">Vectores y Valores propios</a>
        <ol type="i">
            <li><a href="#Rank">RankPage</a></li>
            <ol type="a">
            <li><a href="#a">Inciso a</a></li>
            <li><a href="#b">Inciso b</a></li>
            <li><a href="#c">Inciso c</a></li>
            <li><a href="#d">Inciso d</a></li>
            </ol>
        </ol>
    </li>
    <li><a href="#4">Modelo de Leslie para la dinámica de poblaciones</a></li>
        <ol type="i">
            <li><a href="#I">Ejercicio I</a></li>
            <li><a href="#II">Ejercicio II</a></li>
            <li><a href="#III">Ejercicio III</a></li>
            <li><a href="#IV">Ejercicio IV</a></li>
            <li><a href="#V">Ejercicio V</a></li>
        </ol>
    <li><a href = "preguntas">Preguntas según el modelo de Leslie</a></li>
        <ol type ="i">
        <li><a href="#VI">Pregunta VI</a></li>
        <li><a href="#VII">Pregunta VII</a></li>
        <li><a href="#VIII">Pregunta VIII</a></li>
        </ol>
    <li><a href="#conclusion">Conclusión</a></li>
</ol>


# __Vectores y valores propios__ <a id="voyva"></a>


Vivimos en la era de las computadoras. Internet forma parte de nuestro día a día y la información está a sólo un clic de distancia. Simplemente abra su motor de búsqueda favorito, como Google, AltaVista, Yahoo, escriba las palabras clave y el motor de búsqueda mostrará las páginas relevantes para su búsqueda. Pero, ¿cómo funciona realmente un motor de búsqueda?


Para este objetivo, comenzamos imaginando la red Web como un grafo dirigido, con los nodos representados por páginas web y los bordes representados por los enlaces entre ellas.


Supongamos, por ejemplo, que tenemos una pequeña red de Internet que consta de sólo 4 sitios web www.page1.com, www.page2.com, www.page3.com, www.page4.com, referenciados entre sí en la forma sugerida en la lista :

<table>
  <tr style="background-color: white">
    <td><form action="#">
      <label for="lang">www.page1.com</label>
      <select name="lenguajes" id="lang">
        <option value="selecciona">Selecciona una página</option>
        <option value="javascript">www.page2.com</option>
        <option value="php">www.page3.com</option>
        <option value="java">www.page4.com</option>
      </select>
</form></td>
    <td><form action="#">
      <label for="lang">www.page2.com</label>
      <select name="lenguajes" id="lang">
        <option value="selecciona">Selecciona una página</option>
        <option value="javascript">www.page3.com</option>
        <option value="php">www.page4.com</option>
      </select>
</form></td>
    <td><form action="#">
      <label for="lang">www.page3.com</label>
      <select name="lenguajes" id="lang">
        <option value="selecciona">Selecciona una página</option>
        <option value="javascript">www.page1.com</option>
      </select>
</form></td>
    <td><form action="#">
      <label for="lang">www.page4.com</label>
      <select name="lenguajes" id="lang">
        <option value="selecciona">Selecciona una página</option>
        <option value="javascript">www.page3.com</option>
        <option value="javascript">www.page1.com</option>
      </select></td>
  </tr>
</table>

# __RankPage__<a id="Rank"></a>

<a id="a"></a>
> __a) Traduzca la lista combinada a  un grafo dirigido con 4 nodos, uno para cada sitio web. Cuando el sitio web i hace referencia a j , agregue un borde dirigido entre el nodo i y el nodo j en el grafo. Con el fin de calcular el ranking de su página, ignore cualquier enlace de navegación, como los botones Atrás y Siguiente, solo tome encuenta las conexiones entre diferentes sitios web.__


<center><img src="nodos.png" alt=""></center>


<a id="b"></a>
> __b) En el modelo, cada página debe transferir uniformemente su importancia a las páginas a las que enlaza. Denote por $x_1$ , $x_2$ , $x_3$ y $x_4$ la importancia de las cuatro páginas. Analice la situación en cada nodo y obtenga un sistema lineal, represente su sistema en la forma $Ax=x$__

$$
\begin{align}
    X_1 &= X_3 + \frac{1}{2}X_4 \\ \\
    X_2 &= \frac{1}{3}X_1 \\ \\
    X_3 &= \frac{1}{3}X_1 + \frac{1}{2}X_2 + \frac{1}{2}X_4 \\ \\
    X_4 &= \frac{1}{3}X_1 + \frac{1}{2}X_2
\end{align}
$$

La matriz resultante es : 
$$
A = \begin{pmatrix}
0 & 0 & 1 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & 0 & 0 & 0 \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & 0 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & 0 & 0
\end{pmatrix}
$$

Dándole la forma $Ax = x $ :
$$
\begin{bmatrix}
0 & 0 & 1 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & 0 & 0 & 0 \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & 0 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & 0 & 0
\end{bmatrix}
\begin{bmatrix}
   X_1 \\
   X_2\\
   X_3\\
   X_4 
\end{bmatrix}
= 
\begin{bmatrix}
   X_1 \\
   X_2\\
   X_3\\
   X_4 
\end{bmatrix}
$$


<a id="c"></a>
> __c) Hallar el vector propio de la matriz $A$ asociado al valor propio $\lambda=1$ ¿Cuál es la razón de seleccionar $\lambda=1$?__

+ __Para__ $λ=1$
$(A-\lambda I)V = 0$ :
$$
\begin{align}
\begin{pmatrix}
-1 & 0 & 1 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & -1 & 0 & 0 \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & -1 & \frac{1}{2} \\
\frac{1}{3\vphantom{1/2}} & \frac{1}{2} & 0 & -1
\end{pmatrix}
\begin{pmatrix}
   X_1 \\
   X_2\\
   X_3\\
   X_4 
\end{pmatrix}
=
\begin{pmatrix}
   0\\
   0\\
   0\\
   0
\end{pmatrix}
\end{align}
$$


Obtenemos el siguiente sistema de ecuaciones : 
$$
\begin{align}
    &-x_1 + x_3 + \frac{1}{2}x_4 = 0\\
    &\frac{1}{3}x_1 - x_2 = 0\\
    &\frac{1}{3}x_1 + \frac{1}{2}x_2 -x_3 + \frac{1}{2}x_4= 0\\
    &\frac{1}{3}x_1 + \frac{1}{2}x_2 -x_4 = 0\\
\end{align}
$$


Resolviendo el sistema de ecuaciones obtenemos : 
$$
\begin{align}
   x_1=\frac{4x_3}{3},\:x_4=\frac{2x_3}{3},\:x_2=\frac{4x_3}{9}
\end{align}
$$

Por lo tanto el vector propio para $λ=1$ es cualquier vector de la forma : 
$$
V =\begin{pmatrix}
   \frac{4}{3\vphantom{1/2}}x_3\\
   \frac{4\vphantom{1/2}}{9\vphantom{1/2}}x_3 \\
    x_3\\
   \frac{2\vphantom{1/2}}{3}x_3
\end{pmatrix}
$$
Factorizando tenemos : 
$$
V =\begin{pmatrix}
   \frac{4}{3\vphantom{1/2}}\\
   \frac{4\vphantom{1/2}}{9\vphantom{1/2}} \\
    1\\
   \frac{2\vphantom{1/2}}{3}
\end{pmatrix}
$$

<a id="d"></a>
> __d) Suponga que inicialmente la importancia se distribuye uniformemente entre los 4 nodos y cada uno obtiene ¼. Denota por $x_0$ el vector de rango inicial, en el que todas las entradas son iguales a ¼. construya una función que tenga como entradas la matrix A, el vector inicial $x_0$ y el número de iteraciones. Use la fución para hallar el vector propio dominante (vector PageRank de del grafo web). Mencione la página de mayor relevancia. ¿Qué concluye al tener conocimiento de  la página de mayor relevancia?__

In [1]:
import numpy as np  
from IPython.display import display, Math
"""
Metodo de la potencia
x como vector
a como matriz
"""
def power_method(x,a,num_iter=20):
    resultados = []
    for i in range(num_iter):
        x = np.dot(a,x)
        resultados.append(x)
    return np.array(resultados)


In [2]:
x = np.array([1/4,1/4,1/4,1/4])
a = np.array([[0, 0, 1, 1/2],[1/3, 0, 0, 0],[1/3, 1/2, 0, 1/2],[1/3, 1/2, 0, 0]])
vector_page_rank = power_method(x,a)
vector_page_rank[-1]


array([0.38709616, 0.12903205, 0.29032302, 0.19354877])

El vector anterior es denominado vector PageRank del grafo web.
$$
\begin{align*}
    x = \begin{bmatrix}
        0.3870\\
        0.1290\\
        0.2903 \\
        0.1935 
    \end{bmatrix}
\end{align*}
$$
El vector de PageRank que se ha calculado indica que la página 1 es la más relevante. Esto puede resultar sorprendente, ya que la página 1 tiene solo 2 enlaces entrantes, mientras que la página 3 tiene 3 enlaces entrantes. Al examinar el grafo, se observa  que el nodo 3 tiene únicamente un enlace saliente hacia el nodo 1, transfiriendo así toda su importancia a este último. De manera equivalente, un usuario de internet que sigue únicamente hipervínculos, al llegar a la página 3, solo puede dirigirse a la página 1.



<center><img src="nodos.png" alt=""></center>

## __EL MODELO DE LESLIE PARA LA DINÁMICA DE POBLACIONES.__ <a id="Leslie"></a>




Uno de los modelos que utilizan los demógrafos para predecir el crecimiento de una población es el modelo de Leslie. Este modelo se desarrolló alrededor de 1940.

El modelo describe el crecimiento numérico de una población humana, animal o vegetal, la cuál es dividida en clases que se definen tomando en cuenta rangos de edad, peso, altura, u otros parámetros.
Dividiendo la población inicial en $n$ clases, se forma un vector columna, llamado vector de la distribución inicial de clase $x^{(0)}$ es igual

\begin{pmatrix} x_{1}^{(0)}\\ v_{2}^{(0)} \\ \vdots \\ x_{n}^{(0)} \end{pmatrix}

donde los $x^{(0)}_i$ $i=1,2,...,n$ son el número de elementos en cada clase, en un tiempo inicial $t=0$.


Conforme transcurre el tiempo y debido a los procesos naturales (nacimiento, envegecimiento y muerte), cambia el número de elementos de cada una de las $n$ clases. Describiendo los procesos naturales cuantitativamente, el vector de la distribución inicial de clases puede proyectarse hacia el futuro.

Denotemos al vector de la distribución de clases en cualquier tiempo $t_k$ como:
$x^{k}$ igual

$$
\begin{pmatrix} x_{1}^{(k)}\\ v_{2}^{(k)} \\ \vdots \\ x_{n}^{(k)} \end{pmatrix} \
$$

donde los $x^{k}_i$ son el número de elementos de la clase de orden $i$ en el tiempo $t_k$.

Los procesos naturales pueden describirse con los siguientes parámetros:

* a) $a_i$ Es el número promedio de hijas que tiene una hembra de la clase de orden $i$.

* b) $b_i$ Es la probabilidad de que un elemento de la clase de orden $i$ sobreviva y pase a la clase siguiente $(i+1)$.

El inciso a) y b) se interpreta matemáticamente como
 
* a) $a_i\geq 0$	para $i=1,2,...,n$

* b) $0  ≤ b_i ≤ 1$ para $i=1,2,...,n-1$

* __Represente en lenguaje algebraico lo siguiente:__ 

<a id="I"></a>
> __I) En el tiempo $t_k$ , el número de elementos de la primera clase, serán únicamente los elementos nacidos entre los tiempos $t_{k-1}$ y $t_{k}$.__

$$
\begin{align}
    x_1(k) = a_1x_1(k-1)+a_2x_2(k-1) + \cdots + a_nx_n(k-1)
\end{align}
$$

<a id="II"></a>
> __II) Ahora, el número de elementos de la clase de orden $(i+1)$, en el tiempo $t_k$ , es igual al número de elementos de la clase de orden $i$ en el tiempo $t_{k-1}$, que sobreviven en el tiempo $t_k$.__

$$
\begin{align}
    x_{1+1}(k) = b_ix_i(k-1), \quad i = 1,2, \cdots, n-1
\end{align}
$$

<a id="III"></a>
> __III) Utilizando la notación matricial, escriba i) y ii) en forma matricial, la matriz que obtiene se llama matriz de Leslie.__

$$
\begin{pmatrix}
x_1(k)\\
x_2(k)\\
\vdots \\
x_n(k)
\end{pmatrix}
= 
\begin{pmatrix}
 a_1  & a_2 & \cdots & a_{n-1} & a_n \\
 b_1  & 0 & \cdots & 0 & 0 \\
 0 & b_2 & \cdots & 0 & 0 \\
 \vdots  & \vdots & \vdots & \vdots & \vdots \\
 0 & 0 & \cdots  &b_{n-1} & 0 \\
\end{pmatrix}
\begin{pmatrix}
x_1(k-1)\\
x_2(k-1)\\
x_3(k-1)\\
\vdots \\
x_n(k-1)
\end{pmatrix}
$$

<a id="IV"></a>
> __IV) Considérese una población animal (exclusiva mente hembras) ficticia que consiste de 1000 animales con edad entre 0 y 1, 800 animales con edad entre 1 y 2  y  600 con edad entre 2 y 3. Se supone que ninguno de los animales vive más de tres años.__


$$
x_0 = \begin{pmatrix}
1000 \\
800 \\
600
\end{pmatrix}
$$

<a id="V"></a>
> __V) Considere lo siguiente:La mitad de los animales de la clase 1 sobrevivirán para estar en la clase 2 el próximo año, la mitad de los animales de la clase 2 sobrevivirán para estar en la clase 3 el próximo año. Los animales de la clase 1 no producen hijos, cada animal de la clase 2 produce 1 hijo en promedio, y cada animal de la clase 3 produce 2 hijos en promedio. Con estos datos se forme la matriz de Leslie.__


Organizamos los datos del ejercicio en la siguiente tabla : 

$$
\begin{array}{|c|c|c|c|}
\hline
\text{Clase} & \text{Numero de Hembras} & \text{Natalidad}& \text{Tasa de supervivencia} \\
\hline
\text{1} & 100 & \text{0} &\text{0.5}\\
\text{2} & 800 & \text{1} & \text{0.5} \\
\text{3} & 600 & \text{2} & \text{1} \\
\hline
\end{array}
$$
Recordemos que : 
* $a_i :\quad \text{representa la tasa de natalidad de la clase de edad }i$
* $b_i :\quad \text{representa la tasa de supervivencia de la clase de edad }i\text{ a la siguiente clase de edad }i+1$


Con esto sabido rellenamos la matriz con los datos de la tabla para obtener la matriz de Leslie:

$$
A = \begin{pmatrix}
  0 & 1 & 2 \\
  0.5 & 0 & 0 \\
  0 & 0.5 & 0
\end{pmatrix}
$$

## __Basados en IV), V) y usando el método de potencia responda las siguientes preguntas__
<a id="preguntas"></a>

<a id="VI"></a>
> __VI) ¿Cuál es la distribución én las clases después de un año?__


Pues para calcular la distribución después de un año usando el modelo de Leslie hacemos: 

$$
\begin{align}
    x^{(1)} = A \cdot x^{(0)}
\end{align}
$$
Entonces se puede calcular haciendo : 

$$
X^{(1)} = \begin{pmatrix}
  0 & 1 & 2 \\
  0.5 & 0 & 0 \\
  0 & 0.5 & 0
\end{pmatrix}
\begin{pmatrix}
1000 \\
800 \\
600
\end{pmatrix}
$$

In [3]:
v = np.array([1000, 800, 600])
A = np.array([[0, 1, 2],
              [0.5, 0, 0],
              [0, 0.5, 0]])
x1 = np.dot(A,v)
display(Math(r'x_1 = ' +  np.array2string(x1, separator=', ')))

<IPython.core.display.Math object>

La distribución después de 1 años es : 
$$
X^{(1)} = \begin{pmatrix}
2000 \\
500 \\
400
\end{pmatrix}
$$

__Usando el método de la potencia :__ 

In [4]:
x_1 = power_method(v, A,1) #llamamos al metodo de la potencia
for i, clase in enumerate(v):
    print(f"Clase {i+1}: {x_1[0][i]}")
display(Math(r'x_1 = ' +  np.array2string(x_1, separator=', ')))

Clase 1: 2000.0
Clase 2: 500.0
Clase 3: 400.0


<IPython.core.display.Math object>

`La distribución después de 1 años usando el método de la potencia es :` 

$$
X^{(1)} = \begin{pmatrix}
2000 \\
500 \\
400
\end{pmatrix}
$$

<a id="VII"></a>
> __VII) ¿Cuál es la distribución de edades después de 2 años?__


Usando el método de Leslie : 

In [5]:
x2 = np.dot(A,x1)
display(Math(r'x_2 = ' +  np.array2string(x2, separator=', ')))

<IPython.core.display.Math object>

La distribución después de 2 años es : 
$$
X^{(2)} = \begin{pmatrix}
1300 \\
1000 \\
250
\end{pmatrix}
$$

__Usando el método de la potencia :__

In [6]:
x_2 = power_method(v, A,2) #llamamos al metodo de la potencia
for i, clase in enumerate(v):
    print(f"Clase {i+1}: {x_2[1][i]} ")
display(Math(r'x_2 = ' +  np.array2string(x_2, separator=', ')))

Clase 1: 1300.0 
Clase 2: 1000.0 
Clase 3: 250.0 


<IPython.core.display.Math object>

`La distribución después de 2 años usando el método de la potencia es :`

$$
X^{(2)} = \begin{pmatrix}
1300 \\
1000 \\
250
\end{pmatrix}
$$

<a id="VIII"></a>
> __VIII)¿Qué se puede decir acerca del vector de distribución__ $x^{(k)}$? __suponga__ $k=64$


In [7]:
x_64 = power_method(v,A,64)
display(Math(r'x_{64} = ' +  np.array2string(x_64[63], separator=', '))) # imprimimos solo el resultado

#descomente la linea siguiente si desea ver todas las iteraciones : 
#x_64

<IPython.core.display.Math object>

La distribución de $x^{(k)}$ cuando $k=64$ es :
$$
X^{(64)} = \begin{pmatrix}
1519.999 \\
760.000 \\
380.000
\end{pmatrix}
$$

* El vector de distribución $ x^{(k)} $ nos proporciona información sobre cómo evoluciona la estructura de edades de la población después de $ k $ períodos de tiempo, en este caso, $ k = 64 $ años. Este vector representa el número de individuos en cada clase de edad después de transcurrir 64 años, considerando las tasas de natalidad y supervivencia definidas por el modelo de Leslie. Es una herramienta fundamental para comprender cómo cambia la composición de la población a lo largo del tiempo y cómo factores como la fertilidad y la mortalidad afectan su distribución por edad.

# __Conclusión__ <a id="conclusion"></a>

<p style="text-align: justify;">
En conclusión, el cálculo de los autovalores y autovectores nos ha permitido comprender el comportamiento de sistemas de ecuaciones, tanto lineales como no lineales, así como entender su estructura subyacente. Del mismo modo, el estudio del modelo de Leslie nos ha proporcionado una comprensión más profunda de los cambios en la población en respuesta a factores como la natalidad y la supervivencia. A través de la aplicación de sistemas de ecuaciones, hemos podido predecir el comportamiento de la población a lo largo del tiempo. Estas herramientas nos brindan una sólida base para analizar y comprender dinámicas complejas en diversos campos. 
</p>
<p style="text-align: justify;">
Según todo esto, pienso que es un paso más hacia el acercamiento a las ecuaciones diferenciales y posiblemente el estudio del modelo Leslie de anterior pueda permitirnos utilizar en algún momento el Sistema depredador-presa de ecuaciones diferenciales que, al igual que el modelo de Leslie, toma en cuenta la natalidad, pero este estudia también la interacción entre un depredador y una presa y cómo influye en su población respecto al tiempo.
</p>



# [inicio](#inicio)